File: cpp/ParkingLot.cpp
#include "ParkingLot.hpp"
#include <iostream>
#include <algorithm>

// Constructor
ParkingLot::ParkingLot(int motorcycleCapacity, int carCapacity)
    : motorcycleCapacity(motorcycleCapacity), carCapacity(carCapacity) {
}

// Destructor
ParkingLot::~ParkingLot() {
    // Destructor content, if needed
}

int ParkingLot::searchCarPos(std::string numberPlate) {
    bool found = false;
    int itr = 0, size = listOfCar.size();
    while(itr < size || !found) {
        if(listOfCar[itr].getNumberPlate() != numberPlate) {
            itr++;
        } else {
            found = true;
        }
    }

    return itr;
}

int ParkingLot::searchMotorcyclePos(std::string numberPlate) {
    bool found = false;
    int itr = 0, size = listOfMotorcycle.size();
    while(itr < size || !found) {
        if(listOfMotorcycle[itr].getNumberPlate() != numberPlate) {
            itr++;
        } else {
            found = true;
        }
    }

    return itr;
}

// Getters
int ParkingLot::getMotorcycleCapacity() const {
    return motorcycleCapacity;
}

int ParkingLot::getCarCapacity() const {
    return carCapacity;
}

std::vector<Car>& ParkingLot::getCars() {
    return listOfCar;
}

std::vector<Motorcycle>& ParkingLot::getMotorcycles() {
    return listOfMotorcycle;
}

// Setters
void ParkingLot::setMotorcycleCapacity(int capacity) {
    motorcycleCapacity = capacity;
}

void ParkingLot::setCarCapacity(int capacity) {
    carCapacity = capacity;
}

void ParkingLot::addCar(const Car &car) {
    if(listOfCar.size() < carCapacity) listOfCar.push_back(car);
    else std::cout << "There is not parking lot for" << car.getNumberPlate() << "\n";
}

void ParkingLot::addMotorcycle(const Motorcycle &motorcycle) {
    if(listOfMotorcycle.size() < motorcycleCapacity) listOfMotorcycle.push_back(motorcycle);
    else std::cout << "There is not parking lot for " << motorcycle.getNumberPlate() << "\n";
}

void ParkingLot::startParkingTheCar(const Car &car) {
    addCar(car);
}

void ParkingLot::startParkingTheMotorcycle(const Motorcycle &motorcycle) {
    addMotorcycle(motorcycle);
}

void ParkingLot::finishParkingTheCar(const Car &car) {
    int pos = searchCarPos(car.getNumberPlate());
    if(pos < listOfCar.size()) {
        std::vector<Car>::iterator locationOfCar = listOfCar.begin() + pos;
        listOfCar.erase(locationOfCar);
    } else {
        std::cout << "Number of plate doesn't match\n";
    }
}

void ParkingLot::finishParkingTheMotorcycle(const Motorcycle &motorcycle) {
    int pos = searchMotorcyclePos(motorcycle.getNumberPlate());
    if(pos < listOfMotorcycle.size()) {
        std::vector<Motorcycle>::iterator locationOfMotorcycle = listOfMotorcycle.begin() + pos;
        listOfMotorcycle.erase(locationOfMotorcycle);
    } else {
        std::cout << "Number of plate doesn't match\n";
    }
}

void ParkingLot::print() {
    std::cout << "Parking Lot" << std::endl;

    std::cout << "\nList of number plate of car\n";
    for(Car car: listOfCar) {
        std::cout << car.getNumberPlate() << std::endl;
    }

    std::cout << "\nList of number plate of motorcyle\n";
    for(Motorcycle motorcycle: listOfMotorcycle) {
        std::cout << motorcycle.getNumberPlate() << std::endl;
    }

    std::cout << std::endl;
}

File: php/ParkingLot.php
<?php

require_once 'Car.php';
require_once 'Motorcycle.php';
require_once "Vehicle.php";

class ParkingLot {
    private $motorcycleCapacity;
    private $carCapacity;
    private $listOfCar;
    private $listOfMotorcycle;

    // Constructor
    public function __construct($motorcycleCapacity, $carCapacity) {
        $this->motorcycleCapacity = $motorcycleCapacity;
        $this->carCapacity = $carCapacity;
        $this->listOfCar = array();
        $this->listOfMotorcycle = array();
    }

    // Getters
    public function getMotorcycleCapacity() {
        return $this->motorcycleCapacity;
    }

    public function getCarCapacity() {
        return $this->carCapacity;
    }

    public function getCars() {
        return $this->listOfCar;
    }

    public function getMotorcycles() {
        return $this->listOfMotorcycle;
    }

    // Setters
    public function setMotorcycleCapacity($capacity) {
        $this->motorcycleCapacity = $capacity;
    }

    public function setCarCapacity($capacity) {
        $this->carCapacity = $capacity;
    }

    // Parking methods
    public function addCar($car) {
        if (count($this->listOfCar) < $this->carCapacity) {
            $this->listOfCar[] = $car;
        } else {
            echo "There is no parking lot for " . $car->getNumberPlate() . "<br/>";
        }
    }

    public function addMotorcycle($motorcycle) {
        if (count($this->listOfMotorcycle) < $this->motorcycleCapacity) {
            $this->listOfMotorcycle[] = $motorcycle;
        } else {
            echo "There is no parking lot for " . $motorcycle->getNumberPlate() . "<br/>";
        }
    }

    public function startParkingTheCar($car) {
        $this->addCar($car);
    }

    public function startParkingTheMotorcycle($motorcycle) {
        $this->addMotorcycle($motorcycle);
    }

    public function finishParkingTheCar($car) {
        $pos = $this->searchCarPos($car->getNumberPlate());
        if ($pos !== -1) {
            array_splice($this->listOfCar, $pos, 1);
        } else {
            echo "Number plate doesn't match<br/>";
        }
    }

    public function finishParkingTheMotorcycle($motorcycle) {
        $pos = $this->searchMotorcyclePos($motorcycle->getNumberPlate());
        if ($pos !== -1) {
            array_splice($this->listOfMotorcycle, $pos, 1);
        } else {
            echo "Number plate doesn't match<br/>";
        }
    }

    private function searchCarPos($numberPlate) {
        foreach ($this->listOfCar as $key => $car) {
            if ($car->getNumberPlate() === $numberPlate) {
                return $key;
            }
        }
        return -1;
    }

    private function searchMotorcyclePos($numberPlate) {
        foreach ($this->listOfMotorcycle as $key => $motorcycle) {
            if ($motorcycle->getNumberPlate() === $numberPlate) {
                return $key;
            }
        }
        return -1;
    }

    public function print() {
        echo "Parking Lot<br/>";

        echo "<br/>List of number plates of cars<br/>";
        foreach ($this->listOfCar as $car) {
            echo $car->getNumberPlate() . "<br/>";
        }

        echo "<br/>List of number plates of motorcycles<br/>";
        foreach ($this->listOfMotorcycle as $motorcycle) {
            echo $motorcycle->getNumberPlate() . "<br/>";
        }

        echo "<br/>";
    }
}
